python/fastfec.py

```diff
@@ -1,59 +1,386 @@
+                first_line = True
+            line_cache.last_filename = filename
+            line_cache.last_headers = headers
+            if not first_line:
+                # Format the result and return it (if not a header)
+                queue.put((filename.decode('utf8'), line_result(headers, csv_parse(
+                    line.decode('utf8')), types)))
+                queue.join()
+    return line_callback
+
+
+class LibFastFEC:
+    """Python wrapper for the fastfec library
+
+    Usage:
+
+        Initialize the library:
+
+            fastfec = FastFEC()
+
+        Call a parse method. Options:
+
+        * parse: parse the input file line-by-line
+
+            with open('file.fec', 'rb') as f:
+                for form, line in fastfec.parse(f):
+                    print("Got line:", form, line)
+
+        * parse_as_files: parse the input file into output files
+
+            import os
+            from pathlib import Path
+
+            # Custom open method
+            def open_output_file(filename, *args, **kwargs):
+                filename = os.path.join('custom_output_dir', filename)
+                output_file = Path(filename)
+                output_file.parent.mkdir(exist_ok=True, parents=True)
+                return open(filename, *args, **kwargs)
+
+            with open('file.fec', 'rb') as f:
+                fastfec.parse_as_files(f, open=open_output_file)
+
+        Close the library:
+
+            fastfec.free()
+        """
+
+    def __init__(self):
+        self.__init_lib()
 
         # Initialize
-        self.file_descriptors = {}
-        self.last_filename = None
-        self.last_fd = None
-        self.bytes_written = 0
-        self.buffer_read_fn = BUFFER_READ(
-            make_read_buffer(file_handle))
-        write_callback = self.provide_write_callback(output_directory)
-        self.write_callback_fn = CUSTOM_WRITE(write_callback)
-
-        self.persistent_memory_context = self.libfastfec.newPersistentMemoryContext(
-            1)
-        self.fec_context = self.libfastfec.newFecContext(
-            self.persistent_memory_context, self.buffer_read_fn, BUFFER_SIZE, self.write_callback_fn, BUFFER_SIZE, None, None, None, 0, 1)
-
-    def init_lib(self):
-        # TODO: ensure this works on any platform
-        self.libfastfec = CDLL(os.path.join(SCRIPT_DIR, "../zig-out/lib/libfastfec.dylib"))
-
-        # Lay out arg/res types
+        self.persistent_memory_context = self.libfastfec.newPersistentMemoryContext()
+
+    def parse(self, file_handle):
+        """Parses the input file line-by-line
+
+        Arguments:
+            file_handle -- An input stream for reading a .fec file
+
+        Returns:
+            An iterator that receives the form name and a dictionary
+            object describing each line in the file"""
+        # Set up a queue to be able to yield output
+        queue = Queue()
+        done_processing = object()
```

Comment:
The naming is a little confusing, what's the purpose of this?

Reply:
The done_processing is just something you can push to the queue at the end so it knows when to stop (otherwise it would deadlock forever)